CatAnalyzer/plugins/TtbarSingleLeptonAnalyzer.cc

```diff
@@ -251,51 +272,102 @@ void TtbarSingleLeptonAnalyzer::analyze(const edm::Event& iEvent, const edm::Eve
   // PU Info
   //---------------------------------------------------------------------------
   //---------------------------------------------------------------------------
-  edm::Handle<double> PUWeight;
 
-  iEvent.getByToken(puWeight_, PUWeight);
-  b_PUWeight = *PUWeight;
+  if(isMC_) {
+    edm::Handle<float> PUWeight;
+    
+    iEvent.getByToken(puWeight_, PUWeight);
+    b_PUWeight = *PUWeight;
+  }
+  else b_PUWeight = 1;
 
   //---------------------------------------------------------------------------
   //---------------------------------------------------------------------------
-  // Primary Vertex Info
+  // Generated Particles (For Pythia8)
   //---------------------------------------------------------------------------
   //---------------------------------------------------------------------------
 
-  edm::Handle< std::vector<reco::Vertex> > pvHandle;
-  iEvent.getByToken( pvToken_, pvHandle );
-  reco::Vertex primaryVertex;
-
-  int n_vtxs = 0;
-  // Loop over vertices
-  if (pvHandle->size() != 0) {
-    for (unsigned int i=0; i< pvHandle->size(); i++) {
-     reco::Vertex vtx = pvHandle->at(i);
-     if ( fabs(vtx.z())        < 24 &&
-	  vtx.position().Rho() < 2  &&
-	  vtx.ndof()           > 4  &&
-	  !(vtx.isFake())              ) {
-       n_vtxs ++;
-     }
-    }// for(vertex)
-  } // if(vertex)
-  
-  b_nPV = pvHandle->size();
-  b_nGoodPV = n_vtxs;  
+  bool nGenLep = 0;
 
+  if(TTbarMC_ > 0) {
+
+    edm::Handle<edm::View<reco::GenParticle> > genParticles;
+    iEvent.getByToken(genToken_, genParticles);
+    
+    // Gen Status: http://home.thep.lu.se/~torbjorn/pythia81html/ParticleProperties.html
+    // abs(id) == 6  // t-Quark
+    // abs(id) == 5  // b-Quark
+    // abs(id) == 24 // W-Boson
+    // abs(id) == 15 // Tau
+    // abs(id) == 11 // Electron
+    // abs(id) == 13 // Muon
+    
+    for (unsigned int i = 0; i < genParticles->size(); i++){
+      
+      const reco::Candidate & gp = (*genParticles)[i];
+      int id = gp.pdgId();
+      
+      const reco::Candidate *mom;
+      const reco::Candidate *itmom;
+      const reco::Candidate *mommom;
+      mom = (&gp)->mother();
+
+      int momid = 0;
+      int mommomid = 0;
+
+      if(mom !=0){
+	momid    = mom->pdgId();
+	mommomid = momid;
+
+	itmom = (&gp)->mother();
+	// This loop searches the mother's mom of the particle
+	while (momid == mommomid){
+	  mommom = itmom->mother();      
```

Comment:
I dont think so. Actually I have added the iteration over the first mother. It is necessary in the case of taus.